pool-manager.js

```diff
@@ -0,0 +1,132 @@
+var _ = require('underscore');
+
+function PoolManager (options) {
+  options = options || {};
+
+  this.endpoints = [];
+  this._endpointOffset = 0;
+
+  this.isInPool = options.isInPool || _.constant(true);
+  this.onEndpointReturned = options.onEndpointReturned || _.noop;
+  this.onEndpointRegistered = options.onEndpointRegistered || _.noop;
+}
+
+PoolManager.prototype = {
+  getNextEndpoint: function () {
+    var i, l, offset, endpoint;
+
+    for (i = 0, l = this.endpoints.length; i < l; ++i) {
+      offset = (this._endpointOffset + i) % l;
+      endpoint = this.endpoints[offset];
+
+      if (this.isInPool(endpoint)) {
+        this._endpointOffset = offset + 1;
+        return endpoint;
+      }
+    }
+  },
+  updateEndpoints: function (endpoints) {
+    newEndpoints = endpoints.map(function (info) {
+      return new Endpoint(info);
+    });
+
+    for (i = this.endpoints.length; i--;) {
+      matchingEndpoint = _.findWhere(newEndpoints, { url: this.endpoints[i].url });
+
+      if (matchingEndpoint) { // found a match, remove it from `newEndpoints`, since it's not new
+        newEndpoints = _.without(newEndpoints, matchingEndpoint);
+      } else { // didn't find a match in endpoints, so kill that endpoint
+        this.endpoints.splice(i, 1);
+      }
+    }
+
+    newEndpoints.forEach(function (endpoint) {
+      endpoint.callback = this.onEndpointReturned.bind(this, endpoint);
```

Comment:
Would it create a memory leak if inside the constructor we did this?

```js
this.callback = callback.bind(null, this);
```

I wasn't sure. The garbage collector should be able to remove the object, since the function that references it as a bound param is only referred to by the object itself. But because I wasn't sure how V8 would react, I put it down here so I only create it for new entries to the pool - otherwise every 5 seconds the leak would perpetuate a bit.